ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java

```diff
@@ -0,0 +1,409 @@
+          new TtlUpdateRequestInfo(time.milliseconds(), replica));
+      RouterRequestInfo requestInfo = new RouterRequestInfo(hostname, port, ttlUpdateRequest, replica);
+      requestRegistrationCallback.registerRequestToSend(this, requestInfo);
+      replicaIterator.remove();
+      if (RouterUtils.isRemoteReplica(routerConfig, replica)) {
+        LOGGER.trace("Making request with correlationId {} to a remote replica {} in {} ",
+            ttlUpdateRequest.getCorrelationId(), replica.getDataNodeId(), replica.getDataNodeId().getDatacenterName());
+        routerMetrics.crossColoRequestCount.inc();
+      } else {
+        LOGGER.trace("Making request with correlationId {} to a local replica {} ", ttlUpdateRequest.getCorrelationId(),
+            replica.getDataNodeId());
+      }
+      routerMetrics.getDataNodeBasedMetrics(replica.getDataNodeId()).ttlUpdateRequestRate.mark();
+    }
+  }
+
+  /**
+   * @return a {@link TtlUpdateRequest} to send to a replica.
+   */
+  private TtlUpdateRequest createTtlUpdateRequest() {
+    return new TtlUpdateRequest(NonBlockingRouter.correlationIdGenerator.incrementAndGet(), routerConfig.routerHostname,
+        blobId, expiresAtMs, operationTimeMs);
+  }
+
+  /**
+   * Handles a response for a TTL update operation. It determines whether the request was successful and updates
+   * operation tracker. For the same operation, it is possible that different {@link ServerErrorCode} are received from
+   * different replicas. These error codes are eventually resolved to a single {@link RouterErrorCode}.
+   * @param responseInfo The {@link ResponseInfo} to be handled.
+   * @param ttlUpdateResponse The {@link TtlUpdateResponse} associated with this response.
+   */
+  void handleResponse(ResponseInfo responseInfo, TtlUpdateResponse ttlUpdateResponse) {
+    TtlUpdateRequest ttlUpdateRequest = (TtlUpdateRequest) responseInfo.getRequestInfo().getRequest();
+    TtlUpdateRequestInfo ttlUpdateRequestInfo = ttlUpdateRequestInfos.remove(ttlUpdateRequest.getCorrelationId());
+    // ttlUpdateRequestInfo can be null if this request was timed out before this response is received. No
+    // metric is updated here, as corresponding metrics have been updated when the request was timed out.
+    if (ttlUpdateRequestInfo == null) {
+      return;
+    }
+    ReplicaId replica = ttlUpdateRequestInfo.replica;
+    long requestLatencyMs = time.milliseconds() - ttlUpdateRequestInfo.startTimeMs;
+    routerMetrics.routerRequestLatencyMs.update(requestLatencyMs);
+    routerMetrics.getDataNodeBasedMetrics(replica.getDataNodeId()).ttlUpdateRequestLatencyMs.update(requestLatencyMs);
+    // Check the error code from NetworkClient.
+    if (responseInfo.getError() != null) {
+      LOGGER.debug("TtlUpdateRequest with response correlationId {} timed out for replica {} ",
+          ttlUpdateRequest.getCorrelationId(), replica.getDataNodeId());
+      updateOperationState(replica, RouterErrorCode.OperationTimedOut);
```

Comment:
A responseInfo error always indicates a timeout?

Reply:
right - i think that's how it's been modeled in all the operations